rosbag2_cpp/src/rosbag2_cpp/service_utils.cpp

```diff
@@ -0,0 +1,159 @@
+    std::strlen(service_event_topic_type_postfix),
+    service_event_topic_type_postfix) == 0;
+}
+
+std::string service_event_topic_name_to_service_name(const std::string & topic_name)
+{
+  std::string service_name;
+  if (topic_name.length() <= (sizeof(RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX) - 1)) {
+    return service_name;
+  }
+
+  if (topic_name.substr(
+      topic_name.length() -
+      (sizeof(RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX) - 1)) !=
+    RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX)
+  {
+    return service_name;
+  }
+
+  service_name = topic_name.substr(
+    0, topic_name.length() - (sizeof(RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX) - 1));
+
+  return service_name;
+}
+
+std::string service_event_topic_type_to_service_type(const std::string & topic_type)
+{
+  std::string service_type;
+  if (topic_type.length() <= std::strlen(service_event_topic_type_postfix)) {
+    return service_type;
+  }
+
+  // Should include '/srv/' in type
+  if (topic_type.find(service_event_topic_type_middle) == std::string::npos) {
+    return service_type;
+  }
+
+  if (topic_type.substr(topic_type.length() - std::strlen(service_event_topic_type_postfix)) !=
+    service_event_topic_type_postfix)
+  {
+    return service_type;
+  }
+
+  service_type = topic_type.substr(
+    0, topic_type.length() - strlen(service_event_topic_type_postfix));
+
+  return service_type;
+}
+
+size_t get_serialization_size_for_service_metadata_event()
+{
+  // Since the size is fixed, it only needs to be calculated once.
+  static size_t size = 0;
+
+  if (size != 0) {
+    return size;
+  }
+
+  const rosidl_message_type_support_t * type_support_info =
+    rosidl_typesupport_cpp::get_message_type_support_handle<service_msgs::msg::ServiceEventInfo>();
+
+  // Get the serialized size of service event info
+  const rosidl_message_type_support_t * type_support_handle =
+    rosidl_typesupport_cpp::get_message_typesupport_handle_function(
+    type_support_info,
+    rosidl_typesupport_introspection_cpp::typesupport_identifier);
+  if (type_support_handle == nullptr) {
+    throw std::runtime_error("Cannot get ServiceEventInfo typesupport handle !");
+  }
+
+  auto service_event_info =
+    static_cast<const rosidl_typesupport_introspection_cpp::MessageMembers *>(
+    type_support_handle->data);
+
+  // endian type (4 size) + service event info size + empty request (4 bytes)
+  // + emtpy response (4 bytes)
+  size = 4 + service_event_info->size_of_ + 4 + 4;
+
+  return size;
+}
+
+std::string service_name_to_service_event_topic_name(const std::string & service_name)
+{
+  if (service_name.empty()) {
+    return service_name;
+  }
+
+  // If the end of string is RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX, do nothing
+  if ((service_name.length() > sizeof(RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX)) &&
+    (service_name.substr(service_name.length() - strlen(RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX)) ==
+    RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX))
```

Comment:
@Barry-Xu-2018 And we have finally been caught on `+1` problem with `sizeof(RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX)`. It should be `sizeof(RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX) - 1`. Since sizeof includes NULL terminator at the end of the string literals.

To avoid similar mistakes in the future (or maybe I have overlooked even more) could you please search all over the rosbag2 code and replace all occurrences of the `sizeof(RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX) - 1` and `sizeof(RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX)` by **strlen(RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX)** ?

Reply:
Well after more close consideration this is not really bug since there not strict comparison without equal sign. 
However it would be safer to use strlen() instead of the sizeof()

Reply:
> sizeof(RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX). It should be sizeof(RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX) - 1. Since sizeof includes NULL terminator at the end of the string literals.
> To avoid similar mistakes in the future (or maybe I have overlooked even more) could you please search all over the rosbag2 code and replace all occurrences of the sizeof(RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX) - 1 and sizeof(RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX) by strlen(RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX) ?

Yes. Using strlen can make the code clearer.
I will search all changed code in rosbag2 to replace sizeof by strlen.